### DISCUSSION TITLE
how do i install local library "tutorials" in the visual studio code

### DISCUSSION THREAD

**Discussion by chinnarinakkella** [acturtle/cashflower#97] [General]
when i am running the code,i am getting the error at "from tutorials.annuity.input import policy, assumption" module not found : tutorial...i want to install the library...

**Comment by zchmielewska** [comment]
Hi @chinnarinakkella,

if you want to run the model from the tutorials folder in Visual Studio Code, then you need to:

1. Change in the `input.py` script:

from:
```python
assumption["mortality"] = pd.read_csv("./input/mortality.csv")
assumption["interest_rates"] = pd.read_csv("./input/interest_rates.csv")
```

to:
```python
assumption["mortality"] = pd.read_csv("tutorials/annuity/input/mortality.csv")
assumption["interest_rates"] = pd.read_csv("tutorials/annuity/input/interest_rates.csv")
```

2. Run the model in the terminal using the following command:

```
python -m tutorials.annuity.run
```

Please let me know if that helped or if you have any other question!

P.S. The models in the tutorials folder are there to support the section in [the documentation](https://cashflower.acturtle.com/tutorials.html).

**Reply by chinnarinakkella** [reply]
yah..its helpful @zchmielewska...but i am not getting the output...
![Screenshot (3)](https://user-images.githubusercontent.com/108986395/206362800-5f44f4d1-ca01-4530-8b80-55781e3f7d2a.png)

**Reply by chinnarinakkella** [reply]
And also getting error when executing the setup.py
![Screenshot (4)](https://user-images.githubusercontent.com/108986395/206378361-a94b96c9-6807-4aba-a30b-d9c748aa116e.png)

**Comment by zchmielewska** [comment]
Hi @chinnarinakkella,

I do not know exactly what causes your issue and I can't reproduce it.

However, I have two further ideas that we can try out.

**Idea 1:**

Install `setuptools` and the packages from `requirements.txt`. 

terminal
```
pip install setuptools
pip install -r requirements.txt
```

Perhaps the package tries to install from the local folder and you are missing some packages.


**Idea 2:**

You are trying to run from the same folder where the whole package is stored. 
Maybe it will be cleaner to run it from a separate folder where you will have only the annuity model.
I have recorded how to do it and uploaded the video on YouTube.

https://www.youtube.com/watch?v=MDSWMa5UbBw


Hope that helps. Let me know in case of any other questions!

**Reply by zchmielewska** [reply]
Also, to install a package locally, you can use command:

```
python setup.py install
```

**Reply by chinnarinakkella** [reply]
yes @zchmielewska .Its helpful. Thank you

**Reply by zchmielewska** [reply]
Good to hear! :)

**Reply by chinnarinakkella** [reply]
Is there any explanation for the whole code  video of this repository @zchmielewska...

**Reply by zchmielewska** [reply]
@chinnarinakkella, 

There is a short video on the `cashflower` package [here](https://www.youtube.com/watch?v=xuZaymWsUzw&ab_channel=acturtle).

There is also a written documentation on the package, including the tutorials:
[cashflower.acturtle.com](https://cashflower.acturtle.com/annuity.html)

What are you looking for in particular? A walk-through video for the annuity product?

**Reply by chinnarinakkella** [reply]
i am looking for the cashflower.py code explanation @zchmielewska

**Reply by zchmielewska** [reply]
@chinnarinakkella, that's generally the underlying code that enables to evaluate formulas and create reports. What's your project / goal?

**Reply by chinnarinakkella** [reply]
I have an insurance policy data and I want to create the project repository for it.. Based on that I want to understand the whole code....

**Comment by zchmielewska** [comment]
Hi @chinnarinakkella,

below the general description of the whole code. Hope that helps.

The code in the cashflower package helps to transform the formulas defined by the user into the output reports.


**admin.py**


As a first step, the user needs to create a model. The code in the admin.py script helps with this step. The create_model() function copies template files from the model_tpl folder and replaces parts that require a model name.


**start.py**


Once the user has populated files with the formulas, the model gets started using code in the start.py script. The main function is start() . 

The start() function gathers all components of the model:
- settings from settings.py , 
- model points and runplan from input.py ,  
- model variables and constants from model.py .

After that, it creates the instance of the Model class which is used to calculate and save the results.


**cashflow.py**

The cashflow.py script contains the classes for all the main objects, such as:
- Runplan,
- ModelPoint,
- ModelVariable,
- Constant,
- Model.

The model object that has been created in the start.py script creates a queue of model components (variables and constants) in a way that there are no calculation conflicts. After that, it iterates over each policy and calculates results for all of the components. 

After the calculation is complete, the results are saved to the flat files (csv).

Hope that helps as a start. Let me know if you have any other questions!

**Reply by chinnarinakkella** [reply]
1.I want these classes explanation @zchmielewska 
Runplan,
ModelPoint,
ModelVariable,
Constant,
Model.
2. how the path is going from one function to other function and i am unable to know how the t value for all the variables like age. death probability, survival rate is changing and how it is accessing from the table

**Reply by zchmielewska** [reply]
Hi @chinnarinakkella ,

we have simplified a little bit the structure. There is now only Variable (instead of ModelVariable and Constant). 

The general idea is that variable holds the logic which is written in the function's definition. The model gathers model point sets and variables. Then the model goes into the `main` model point set. It calculates all the results for each of the variables. It starts with the first model point, then the second, and so on. At the end, the model gathers all the results and combines them into one output.

The path from one function to other function bases on which function calls which. We start with functions that don't have any dependents, once they are calculated then we go to another and calculate that one. To figure out the order, the `ast` package is used. The `t` is always each integer number between 0 and maximal calculation period (`T_MAX_CALCULATION` in settings).

Do you build a cash flow model? Or what is your project?
